### ISSUE TITLE
RavenDB timeout query can hang endpoint and can hang the shutdown process

### ISSUE THREAD

**Issue by mauroservienti** [Particular/NServiceBus.RavenDB#173]
Task force: @colin-higgins @mauroservienti @DavidBoike

### Symptoms

Endpoint is unable to complete the shutdown procedure due to timeout poller being stuck in `GetNextChunk` query.

### Description

The RavenDB implementation of `GetNextChunk` uses the stream API which can block forever if inserts outpace the reads. In our scenario we are sending out messages (doing I/O operations) inside the loop so if the queue is slower than raven inserts, the query will never return.

Here's a snippet that shows the problem
```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Raven.Client.Document;

namespace RavenNeverendingStream
{
    class Program
    {
        static void Main(string[] args)
        {
            var store = new DocumentStore()
            {
                Url = "http://localhost:8083"
            }.Initialize();

            var producer = Task.Run(() =>
            {
                var i = 0;
                while (true)
                {
                    using (var session = store.OpenSession("Test1"))
                    {
                        session.Store(new Timeout {Number = i});
                        session.SaveChanges();
                        i++;
                    }
                    Thread.Sleep(150);
                }
            });

            var consumer = Task.Run(() =>
            {
                while (true)
                {
                    try
                    {
                        Console.WriteLine("Starting new query.");
                        using (var session = store.OpenSession("Test1"))
                        {
                            Console.WriteLine("Opening stream.");
                            var query = session.Query<Timeout>("SomeIndex");
                            var stream = session.Advanced.Stream(query);
                            Console.WriteLine("Enumerating the stream.");
                            int j = 0;
                            while (stream.MoveNext())
                            {
                                j++;
                                //Console.WriteLine(stream.Current.Document.Id);
                                session.Delete(stream.Current.Document.Id);
                                Thread.Sleep(200);
                            }
                            Console.WriteLine(j);
                        }
                    }
                    catch (Exception ex)
                    {
                        Console.WriteLine(ex);
                    }
                    
                }
            });
            Console.ReadLine();
        }
    }

    public class Timeout
    {
        public string Id { get; set; }
        public int Number { get; set; }
    }
}
```

### Who's affected

RavenDB persistence users who have endpoints with high timeout throughput.

### Suggested resolution

Bound the stream query to not buffer more than a certain number of timeouts

**Comment by mauroservienti** [comment]
This is the same issue that we are handling for [vNext](https://github.com/Particular/NServiceBus.RavenDB/issues/140) here things are a bit more complex. I just checked the code and there are no options to introduce a `FeatureStartupTask`, doesn't exist in V5, so there is no way from the `TimeoutPersister` to know that the endpoint is shutting down.

The only things that comes to my mind is that given that it is a `Singleton` if we implement IDisposable it will be disposed when the bus is disposed, since the container is disposed. Unfortunately not all the containers obey to this rule :cry: 

**In order to bring this in progress we need a proper task force.**

@Particular/persistence-maintainers please chime in.

**Comment by colin-higgins** [comment]
I'd like to take point on this

**Comment by mauroservienti** [comment]
We need a proper taskforce, if no one volunteers I'll join.

**Comment by DavidBoike** [comment]
Can join in a supporting role.

**Comment by colin-higgins** [comment]
The solution we talked about in chat was using `IWantToRunWhenBusStartsAndStops` for v5.
This works because timeouts don't work in a send only configuration.

For v6, the two leading options seemed to be using a cancellation token or having an explicit shutdown delegate registry.

```
public CustomTimeoutPersister(IShutdownDelegateRegistry registry) {
    registry.Add(RunMeWhenEndpointStops);
}
```

**Comment by mauroservienti** [comment]
Yes, it sounds right, let's create a PR for V5 so that we can discuss the implementation.

Regarding V6 there is already an [ongoing discussion](https://github.com/Particular/NServiceBus/issues/3472) and a [PR](https://github.com/Particular/NServiceBus/pull/3497). What is discussed is specifically related to `IQueryTimeouts`, maybe we should open another issue to discuss shutdown in general, that is indeed a different thing compared to cancelling some ongoing operation, or may be not only cancelling an ongoing operation.

**Comment by andreasohlund** [comment]
@colin-higgins @mauroservienti @DavidBoike can we move this one into in-progress?

**Comment by mauroservienti** [comment]
We had a call this morning regarding [the IQueryTimeouts cancellation token](https://github.com/Particular/NServiceBus/pull/3497) and we decided to drop it in favor of https://github.com/Particular/NServiceBus/issues/3532.

So the goal here is to do the minimum amount of required work to prevent the issue to affect customers:

* Use `IWantToRunWhenBusStartsAndStops` to plugin shutdown logic
* Find a way to prevent the streaming API unbounded usage when querying for timouts

**Comment by DavidBoike** [comment]
@mauroservienti The customer who was reliably having hangs discovered that they were doing something funny with reflection on a Task.Result, so they resolved their problem without any intervention from us.

Need input from you on if you think we should ship this as 3.1.1 anyway, make the change for VNext, or leave the stream API as is.

Above, you have a snippet showing "the problem" but it's not NSB code. Are we still susceptible? @colin-higgins and I are loathe to release a "fix" for something that may not even be a problem.

**Comment by mauroservienti** [comment]
As discussed we are not going to ship this fix as we have no evidence that this is affecting anyone.
I removed the milestone and I'll close the issue.

**Comment by DavidBoike** [comment]
Note that #140 was a duplicate of this, and on that one we discovered we are seeing this in the wild, and this will be shipped with 3.0.7, but this was a duplicate so I'm just updating the labels here so there's no confusion.
